Navigation-Examples/Examples/Custom-RoutingProvider.swift

```diff
@@ -0,0 +1,138 @@
+        options.attributeOptions = [.congestionLevel, .speed, .maximumSpeedLimit, .expectedTravelTime]
+    }
+    
+    // Here any manipulations on the reponse data can take place
+    func applyMapMatchingModifications(_ response: MapMatchingResponse) {
+        response.matches?.forEach { match in
+            match.legs.forEach { leg in
+                leg.incidents = fetchExternalIncidents(for: leg)
+            }
+        }
+    }
+    
+    // Let's say we have an external source of incidents data, we want to apply to the route.
+    func fetchExternalIncidents(for leg: RouteLeg) -> [Incident] {
+        return [Incident(identifier: "\(leg.name) incident",
+                         type: .otherNews,
+                         description: "Custom Incident",
+                         creationDate: Date(),
+                         startDate: Date(),
+                         endDate: Date().addingTimeInterval(60),
+                         impact: nil,
+                         subtype: nil,
+                         subtypeDescription: nil,
+                         alertCodes: [],
+                         lanesBlocked: nil,
+                         shapeIndexRange: 0..<1)]
+    }
+    
+    // MARK: RoutingProvider implementation
+    
+    func calculateRoutes(options: RouteOptions, completionHandler: @escaping Directions.RouteCompletionHandler) -> NavigationProviderRequest? {
+        
+        applyOptionsModification(options)
+        
+        // Using `MapboxRoutingProvider` also illustrates cases when we need to modify just a part of the route, or dynamically edit `RouteOptions` for each reroute.
+        return routeCalculator.calculateRoutes(options: options,
+                                               completionHandler: completionHandler)
+    }
+    
+    func calculateRoutes(options: MatchOptions, completionHandler: @escaping Directions.MatchCompletionHandler) -> NavigationProviderRequest? {
+        
+        applyOptionsModification(options)
+        
+        return routeCalculator.calculateRoutes(options: options,
+                                               completionHandler: { [weak self] (session, result) in
+            switch result {
+            case .failure(_):
+                completionHandler(session, result)
+            case .success(let response):
+                guard let strongSelf = self else {
+                    return
+                }
+                strongSelf.applyMapMatchingModifications(response)
+                
+                completionHandler(session, .success(response))
+            }
+        })
+    }
+    
+    // Let's make our custom routing provider prevent route refreshes.
+    func refreshRoute(indexedRouteResponse: IndexedRouteResponse, fromLegAtIndex: UInt32, completionHandler: @escaping Directions.RouteCompletionHandler) -> NavigationProviderRequest? {
+        
+        var options: DirectionsOptions!
```

Comment:
NIT (here and in other methods)
```suggestion
        var options: DirectionsOptions!
```